Import Relevant Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as SNS

from sklearn.dummy import DummyClassifier
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score, GridSearchCV, train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier, GradientBoostingClassifier
from sklearn.metrics import accuracy_score, recall_score, precision_score, confusion_matrix, \
                            plot_confusion_matrix, plot_roc_curve, roc_auc_score
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.neighbors import KNeighborsClassifier

import xgboost

In [2]:
# read in cleaned data file
df = pd.read_csv('final_data.csv', index_col=0)
df.head()

amount_tsh  funder  gps_height installer                    basin  \
0      6000.0   Other        1390     Other               Lake Nyasa   
1         0.0   Other        1399     Other            Lake Victoria   
2        25.0   Other         686     Other                  Pangani   
3         0.0  Unicef         263     Other  Ruvuma / Southern Coast   
4         0.0   Other           0     Other            Lake Victoria   

   region_code  district_code public_meeting scheme_management  permit  \
0           11              5           True               VWC   False   
1           20              2        Unknown             Other    True   
2           21              4           True               VWC    True   
3           90             63           True               VWC    True   
4           18              1           True           Unknown    True   

  extraction_type_class management         payment water_quality  \
0               gravity        vwc    pay annually          soft   
1               gravity        wug       never pay          soft   
2               gravity        vwc  pay per bucket          soft   
3           submersible        vwc       never pay          soft   
4               gravity      other       never pay          soft   

       quantity                source              waterpoint_type  \
0        enough                spring           communal standpipe   
1  insufficient  rainwater harvesting           communal standpipe   
2        enough                   dam  communal standpipe multiple   
3           dry           machine dbh  communal standpipe multiple   
4      seasonal  rainwater harvesting           communal standpipe   

     status_group  age  
0      functional   23  
1      functional   12  
2      functional   13  
3  non functional   36  
4      functional   22

In [3]:
# Create our predictors and target
X = df.drop('status_group', axis=1)
y = df.status_group

In [4]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [5]:
X_train.head()

amount_tsh   funder  gps_height installer                    basin  \
4442          0.0    Other         947     Other  Ruvuma / Southern Coast   
11884         0.0    Other         227     Other  Ruvuma / Southern Coast   
38674      4000.0    Other        1980     Other                   Rufiji   
34584       500.0   Unicef        1471       DWE                   Rufiji   
58645         0.0  Unknown           0   Unknown                   Rufiji   

       region_code  district_code public_meeting scheme_management  permit  \
4442            10              2           True               VWC    True   
11884            8              3           True               VWC   False   
38674           11              7           True               VWC    True   
34584           11              4           True               WUA    True   
58645           12              7           True               VWC    True   

      extraction_type_class management       payment water_quality  quantity  \
4442                gravity        vwc     never pay          soft    enough   
11884               gravity        vwc     never pay       unknown       dry   
38674               gravity        vwc  pay annually          soft    enough   
34584               gravity        wua   pay monthly          soft    enough   
58645               gravity        vwc   pay monthly          soft  seasonal   

                     source     waterpoint_type  age  
4442                  river  communal standpipe   35  
11884  rainwater harvesting  communal standpipe   20  
38674                 river  communal standpipe   19  
34584                 river  communal standpipe   44  
58645                 river  communal standpipe   22

In [6]:
# Confirm that out training data set is 3x larger than the testing dataset
print(X_train.shape, X_test.shape)

(44271, 18) (14758, 18)


# Modeling

### Dummy Model

First, we will create a dummy model that always predicts the most frequent class. In this case, our dummy model will predict that all of the water wells are functional.

In [7]:
# Instantiate our dummy regressor object
dummy = DummyClassifier(strategy='most_frequent')

In [8]:
# Fit the model on our training data
dummy.fit(X_train, y_train)

DummyClassifier(strategy='most_frequent')

In [9]:
# Inspect the accuracy of our initial dummy model
dummy.score(X_train, y_train)

0.5457071220437758

Our dummy model accuracy of 54.6% reflects the distribution of our target. Thus, if we always predict that a water well is functional, we will achieve 54.6% accuracy.

### Set Up Column Transformer

In [10]:
# Create our sub pipelines for numeric and categorial data
subpipe_num = Pipeline(steps=[('ss', StandardScaler())])
subpipe_cat = Pipeline(steps=[('ohe', OneHotEncoder(sparse=False, 
                                                    handle_unknown='ignore'))])

In [11]:
# Include our sub pipelines in a column transformer to OHE our categorial variables and scale our numerical variables.
CT = ColumnTransformer(transformers=[('subpipe_num', subpipe_num, [0,2,17]), 
                                     ('subpipe_cat', subpipe_cat, [1,3,4,5,6,7,8,9,10,11,12,13,14,15,16])])

### Create Pipeline for Simple Models

In [12]:
logreg_pipe = Pipeline(steps=[('CT', CT), 
                              ('logreg', LogisticRegression(random_state=42))])

knn_pipe = Pipeline(steps=[('CT', CT), 
                              ('knn', KNeighborsClassifier())])

dtc_pipe = Pipeline(steps=[('CT', CT), 
                              ('dtc', DecisionTreeClassifier(random_state=42))])


In [13]:
# Fit each pipeline to our training data
logreg_pipe.fit(X_train, y_train)
knn_pipe.fit(X_train, y_train)
dtc_pipe.fit(X_train, y_train)

/Users/zacharypollatsek/opt/anaconda3/envs/learn-env/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Pipeline(steps=[('CT',
                 ColumnTransformer(transformers=[('subpipe_num',
                                                  Pipeline(steps=[('ss',
                                                                   StandardScaler())]),
                                                  [0, 2, 17]),
                                                 ('subpipe_cat',
                                                  Pipeline(steps=[('ohe',
                                                                   OneHotEncoder(handle_unknown='ignore',
                                                                                 sparse=False))]),
                                                  [1, 3, 4, 5, 6, 7, 8, 9, 10,
                                                   11, 12, 13, 14, 15, 16])])),
                ('dtc', DecisionTreeClassifier(random_state=42))])

In [14]:
# Get initial accuracy scores for each our our simple classifier models
logreg_score = logreg_pipe.score(X_train, y_train)
knn_score = knn_pipe.score(X_train, y_train)
dtc_score = dtc_pipe.score(X_train, y_train)

In [19]:
# Print out initial accuracy scores
print(f'Logistic Regression Accuracy: {round(logreg_score * 100, 2)}%')
print(f'KNearestNeighbors Accuracy: {round(knn_score * 100, 2)}%')
print(f'DecisionTree Accuracy: {round(dtc_score * 100, 2)}%')

Logistic Regression Accuracy: 74.68%
KNearestNeighbors Accuracy: 82.61%
DecisionTree Accuracy: 94.4%


Our initial numbers appear to be relatively good. We expect our decision tree classifier to be overfit given the high accuracy score. We will dive deeper into more complex models below to maximize our accuracy and recall. 

### Random Forest

In [23]:
# Create our random forest pipeline to incorporate column transformer
rfc_pipe = Pipeline(steps=[('CT', CT),
                          ('rfc', RandomForestClassifier(random_state=42))])

In [24]:
# Fit our random forest classifier to the training data
rfc_pipe.fit(X_train, y_train)

Pipeline(steps=[('CT',
                 ColumnTransformer(transformers=[('subpipe_num',
                                                  Pipeline(steps=[('ss',
                                                                   StandardScaler())]),
                                                  [0, 2, 17]),
                                                 ('subpipe_cat',
                                                  Pipeline(steps=[('ohe',
                                                                   OneHotEncoder(handle_unknown='ignore',
                                                                                 sparse=False))]),
                                                  [1, 3, 4, 5, 6, 7, 8, 9, 10,
                                                   11, 12, 13, 14, 15, 16])])),
                ('rfc', RandomForestClassifier(random_state=42))])

In [25]:
rfc_pipe.score(X_train, y_train)

0.9439813873641888

In [26]:
cross_val_score(estimator=rfc_pipe, X=X_train, y=y_train, cv=5)

array([0.7796725 , 0.78032528, 0.7798735 , 0.78134177, 0.77716286])

Given that our cross validation scores were significantly lower than the accuracy score from our single random forest classifier, we can say that this model is overfit to the training data. We will perform a grid search in order to optimize the hyper parameters of this classifier.

In [27]:
# create our grid for the grid search parameters
params = {'rfc__n_estimators': [50, 100, 150],
          'rfc__min_samples_split': [2, 10, 50],
          'rfc__max_depth': [5, 10, 15]}

# set up GridSearchCV object
grid_rfc = GridSearchCV(rfc_pipe, param_grid=params, cv=5)

In [28]:
# fit our grid object for random forests to the training data
grid_rfc.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('CT',
                                        ColumnTransformer(transformers=[('subpipe_num',
                                                                         Pipeline(steps=[('ss',
                                                                                          StandardScaler())]),
                                                                         [0, 2,
                                                                          17]),
                                                                        ('subpipe_cat',
                                                                         Pipeline(steps=[('ohe',
                                                                                          OneHotEncoder(handle_unknown='ignore',
                                                                                                        sparse=False))]),
                                            

In [29]:
grid_rfc.best_params_

{'rfc__max_depth': 15, 'rfc__min_samples_split': 2, 'rfc__n_estimators': 100}

In [50]:
grid_rfc.best_score_

0.7832668202933669

From our initial grid search, we attained an accuracy of 78.3% using a max_depth of 15, a min_samples_split of 2, and an n_estimators of 100. 

In [30]:
pd.DataFrame(grid_rfc.cv_results_)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        1.301719      0.011407         0.088801        0.001377   
1        2.387787      0.008930         0.120239        0.002130   
2        3.468344      0.014413         0.152590        0.003440   
3        1.307158      0.018561         0.088471        0.002444   
4        2.397931      0.012328         0.120913        0.002938   
5        3.510307      0.037176         0.151582        0.003910   
6        1.304617      0.006007         0.087595        0.002325   
7        2.385140      0.004705         0.121148        0.002780   
8        3.472620      0.015733         0.153408        0.005454   
9        1.840314      0.006085         0.100718        0.002525   
10       3.474164      0.016416         0.146275        0.002726   
11       5.082416      0.029234         0.191051        0.002289   
12       1.826148      0.007205         0.099117        0.001970   
13       3.493979      0.052523         0.145678        0.003941   
14       5.037199      0.018929         0.188716        0.001371   
15       1.804178      0.016657         0.099236        0.001133   
16       3.406851      0.004665         0.145091        0.002484   
17       5.125888      0.265011         0.198818        0.009734   
18       2.294149      0.021300         0.124621        0.001686   
19       4.391976      0.011384         0.198250        0.002197   
20       6.368161      0.197006         0.268930        0.007237   
21       2.239019      0.023349         0.126586        0.001728   
22       4.291778      0.017242         0.192060        0.001866   
23       6.338778      0.041587         0.260967        0.003281   
24       2.188681      0.026198         0.118381        0.002709   
25       4.179053      0.005387         0.180584        0.002597   
26       6.141256      0.040308         0.244398        0.004725   

   param_rfc__max_depth param_rfc__min_samples_split param_rfc__n_estimators  \
0                     5                            2                      50   
1                     5                            2                     100   
2                     5                            2                     150   
3                     5                           10                      50   
4                     5                           10                     100   
5                     5                           10                     150   
6                     5                           50                      50   
7                     5                           50                     100   
8                     5                           50                     150   
9                    10                            2                      50   
10                   10                            2                     100   
11                   10                            2                     150   
12                   10                           10                      50   
13                   10                           10                     100   
14                   10                           10                     150   
15                   10                           50                      50   
16                   10                           50                     100   
17                   10                           50                     150   
18                   15                            2                      50   
19                   15                            2                     100   
20                   15                            2                     150   
21                   15                           10                      50   
22                   15                           10                     100   
23                   15                           10                     150   
24                   15                           50                      50   
25                 

Based on our initial grid search results, we can infer that a larger max depth is needed to improve accuracy. 

In [31]:
# create our grid for the grid search parameters
params2 = {'rfc__n_estimators': [50, 100, 150],
           'rfc__max_depth': [15, 25, 50]}

# set up GridSearchCV object
grid_rfc2 = GridSearchCV(rfc_pipe, param_grid=params2, cv=5)

In [32]:
# fit our grid object with parameter updates for random forests to the training data
grid_rfc2.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('CT',
                                        ColumnTransformer(transformers=[('subpipe_num',
                                                                         Pipeline(steps=[('ss',
                                                                                          StandardScaler())]),
                                                                         [0, 2,
                                                                          17]),
                                                                        ('subpipe_cat',
                                                                         Pipeline(steps=[('ohe',
                                                                                          OneHotEncoder(handle_unknown='ignore',
                                                                                                        sparse=False))]),
                                            

In [33]:
grid_rfc2.best_params_

{'rfc__max_depth': 25, 'rfc__n_estimators': 150}

In [34]:
grid_rfc2.best_score_

0.7897721682958518

From our second grid search, we attained an accuracy of 79.0% using a max_depth of 25 and an n_estimators of 150. 

In [35]:
pd.DataFrame(grid_rfc2.cv_results_)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0       2.346743      0.038142         0.128413        0.000396   
1       4.451738      0.036551         0.199298        0.002288   
2       6.524388      0.033046         0.269549        0.002430   
3       2.853596      0.019079         0.159120        0.002047   
4       5.488861      0.023299         0.260454        0.004628   
5       8.154395      0.038337         0.362912        0.004045   
6       3.064216      0.014483         0.172049        0.002432   
7       5.992171      0.020339         0.284298        0.000437   
8       8.838696      0.057840         0.400096        0.002814   

  param_rfc__max_depth param_rfc__n_estimators  \
0                   15                      50   
1                   15                     100   
2                   15                     150   
3                   25                      50   
4                   25                     100   
5                   25                     150   
6                   50                      50   
7                   50                     100   
8                   50                     150   

                                             params  split0_test_score  \
0   {'rfc__max_depth': 15, 'rfc__n_estimators': 50}           0.775833   
1  {'rfc__max_depth': 15, 'rfc__n_estimators': 100}           0.777866   
2  {'rfc__max_depth': 15, 'rfc__n_estimators': 150}           0.778543   
3   {'rfc__max_depth': 25, 'rfc__n_estimators': 50}           0.784980   
4  {'rfc__max_depth': 25, 'rfc__n_estimators': 100}           0.785658   
5  {'rfc__max_depth': 25, 'rfc__n_estimators': 150}           0.786110   
6   {'rfc__max_depth': 50, 'rfc__n_estimators': 50}           0.778656   
7  {'rfc__max_depth': 50, 'rfc__n_estimators': 100}           0.780350   
8  {'rfc__max_depth': 50, 'rfc__n_estimators': 150}           0.779447   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0           0.783714           0.784504           0.786311           0.776711   
1           0.785747           0.785182           0.790490           0.777050   
2           0.785521           0.785521           0.788344           0.777276   
3           0.790038           0.788344           0.792862           0.783488   
4           0.791845           0.790603           0.793878           0.784504   
5           0.793314           0.791620           0.793991           0.783827   
6           0.780325           0.779083           0.782358           0.776485   
7           0.780099           0.778292           0.781229           0.775695   
8           0.780890           0.779309           0.781229           0.777615   

   mean_test_score  std_test_score  rank_test_score  
0         0.781415        0.004292                6  
1         0.783267        0.005095                4  
2         0.783041        0.004333                5  
3         0.787943        0.003388                3  
4         0.789298        0.003617                2  
5         0.789772        0.004063                1  
6         0.779382        0.001937                8  
7         0.779133        0.001967                9  
8         0.779698        0.001290                7

Based on our new grid search, we can see that a maximum depth greater than 15 but less than 50 is ideal. We will also try running the grid search with a higher number of trees in the random forest to improve accuracy.

In [36]:
# create our grid for the grid search parameters
params3 = {'rfc__n_estimators': [100, 150, 200],
           'rfc__max_depth': [20, 25, 30]}

# set up GridSearchCV object
grid_rfc3 = GridSearchCV(rfc_pipe, param_grid=params3, cv=5)

In [37]:
# Fit the grid object to our training data with updated grid params for n_estimators and max_depth
grid_rfc3.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('CT',
                                        ColumnTransformer(transformers=[('subpipe_num',
                                                                         Pipeline(steps=[('ss',
                                                                                          StandardScaler())]),
                                                                         [0, 2,
                                                                          17]),
                                                                        ('subpipe_cat',
                                                                         Pipeline(steps=[('ohe',
                                                                                          OneHotEncoder(handle_unknown='ignore',
                                                                                                        sparse=False))]),
                                            

In [39]:
grid_rfc3.best_params_

{'rfc__max_depth': 20, 'rfc__n_estimators': 150}

In [40]:
grid_rfc3.best_score_

0.7932732831246889

From our third grid search, we attained an accuracy of 79.3% using a max_depth of 20 and an n_estimators of 150. 

In [41]:
pd.DataFrame(grid_rfc3.cv_results_)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0       5.114668      0.100837         0.237210        0.003148   
1       7.448913      0.050085         0.330273        0.003286   
2      10.470498      0.729096         0.429496        0.017237   
3       5.773903      0.526419         0.265385        0.003499   
4       8.407452      0.210512         0.376479        0.020376   
5      11.816350      0.881208         0.479841        0.006052   
6       5.945509      0.322162         0.280625        0.002535   
7       9.577850      0.544449         0.414980        0.022117   
8      13.304384      1.550077         0.502973        0.011423   

  param_rfc__max_depth param_rfc__n_estimators  \
0                   20                     100   
1                   20                     150   
2                   20                     200   
3                   25                     100   
4                   25                     150   
5                   25                     200   
6                   30                     100   
7                   30                     150   
8                   30                     200   

                                             params  split0_test_score  \
0  {'rfc__max_depth': 20, 'rfc__n_estimators': 100}           0.789046   
1  {'rfc__max_depth': 20, 'rfc__n_estimators': 150}           0.791756   
2  {'rfc__max_depth': 20, 'rfc__n_estimators': 200}           0.792321   
3  {'rfc__max_depth': 25, 'rfc__n_estimators': 100}           0.785658   
4  {'rfc__max_depth': 25, 'rfc__n_estimators': 150}           0.786110   
5  {'rfc__max_depth': 25, 'rfc__n_estimators': 200}           0.785884   
6  {'rfc__max_depth': 30, 'rfc__n_estimators': 100}           0.780237   
7  {'rfc__max_depth': 30, 'rfc__n_estimators': 150}           0.783060   
8  {'rfc__max_depth': 30, 'rfc__n_estimators': 200}           0.781705   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0           0.795686           0.796476           0.797267           0.786989   
1           0.795008           0.795121           0.797041           0.787441   
2           0.795686           0.794895           0.795686           0.787441   
3           0.791845           0.790603           0.793878           0.784504   
4           0.793314           0.791620           0.793991           0.783827   
5           0.794104           0.792410           0.794556           0.783601   
6           0.784504           0.784504           0.785408           0.778970   
7           0.785747           0.784956           0.785069           0.778292   
8           0.785747           0.785634           0.786085           0.779648   

   mean_test_score  std_test_score  rank_test_score  
0         0.793093        0.004224                3  
1         0.793273        0.003375                1  
2         0.793206        0.003136                2  
3         0.789298        0.003617                6  
4         0.789772        0.004063                5  
5         0.790111        0.004500                4  
6         0.782725        0.002601                9  
7         0.783425        0.002717                8  
8         0.783764        0.002608                7

In [44]:
# create our grid for the grid search parameters
params4 = {'rfc__n_estimators': [100, 150],
           'rfc__max_depth': [18, 20, 22],
           'rfc__max_features': ['auto', 'sqrt', 0.2]}

# set up GridSearchCV object
grid_rfc4 = GridSearchCV(rfc_pipe, param_grid=params4, cv=5)

In [45]:
grid_rfc4.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('CT',
                                        ColumnTransformer(transformers=[('subpipe_num',
                                                                         Pipeline(steps=[('ss',
                                                                                          StandardScaler())]),
                                                                         [0, 2,
                                                                          17]),
                                                                        ('subpipe_cat',
                                                                         Pipeline(steps=[('ohe',
                                                                                          OneHotEncoder(handle_unknown='ignore',
                                                                                                        sparse=False))]),
                                            

In [47]:
grid_rfc4.best_params_

{'rfc__max_depth': 20, 'rfc__max_features': 0.2, 'rfc__n_estimators': 150}

In [48]:
grid_rfc4.best_score_

0.7934991544239145

From our fourth grid search, we attained an accuracy of 79.4% using a max_depth of 20, a max_features of 0.2, and an n_estimators of 150. 

In [49]:
pd.DataFrame(grid_rfc4.cv_results_)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        4.854097      0.079085         0.222644        0.000962   
1        7.026312      0.030303         0.305097        0.001928   
2        4.772689      0.075151         0.224029        0.002694   
3        7.014761      0.042045         0.306500        0.002449   
4        9.448839      0.041567         0.219691        0.001682   
5       13.961023      0.081917         0.303138        0.002018   
6        4.942688      0.019941         0.234814        0.001883   
7        7.298178      0.026983         0.334742        0.007693   
8        4.965089      0.039019         0.238054        0.002290   
9        7.324271      0.083424         0.328797        0.001753   
10       9.687729      0.049160         0.236517        0.002793   
11      14.455038      0.078036         0.322150        0.003063   
12       5.194169      0.083516         0.249187        0.003232   
13       7.786276      0.138943         0.346611        0.002261   
14       5.167387      0.016331         0.250441        0.002582   
15       8.459914      0.223882         0.348941        0.005452   
16      10.418608      0.264010         0.243009        0.005983   
17      16.290158      0.855664         0.354442        0.011671   

   param_rfc__max_depth param_rfc__max_features param_rfc__n_estimators  \
0                    18                    auto                     100   
1                    18                    auto                     150   
2                    18                    sqrt                     100   
3                    18                    sqrt                     150   
4                    18                     0.2                     100   
5                    18                     0.2                     150   
6                    20                    auto                     100   
7                    20                    auto                     150   
8                    20                    sqrt                     100   
9                    20                    sqrt                     150   
10                   20                     0.2                     100   
11                   20                     0.2                     150   
12                   22                    auto                     100   
13                   22                    auto                     150   
14                   22                    sqrt                     100   
15                   22                    sqrt                     150   
16                   22                     0.2                     100   
17                   22                     0.2                     150   

                                               params  split0_test_score  \
0   {'rfc__max_depth': 18, 'rfc__max_features': 'a...           0.788142   
1   {'rfc__max_depth': 18, 'rfc__max_features': 'a...           0.787126   
2   {'rfc__max_depth': 18, 'rfc__max_features': 's...           0.788142   
3   {'rfc__max_depth': 18, 'rfc__max_features': 's...           0.787126   
4   {'rfc__max_depth': 18, 'rfc__max_features': 0....           0.790966   
5   {'rfc__max_depth': 18, 'rfc__max_features': 0....           0.791417   
6   {'rfc__max_depth': 20, 'rfc__max_features': 'a...           0.789046   
7   {'rfc__max_depth': 20, 'rfc__max_features': 'a...           0.791756   
8   {'rfc__max_depth': 20, 'rfc__max_features': 's...           0.789046   
9   {'rfc__max_depth': 20, 'rfc__max_features': 's...           0.791756   
10  {'rfc__max_depth': 20, 'rfc__max_features': 0....           0.790514   
11  {'rfc__max_depth': 20, 'rfc__max_features': 0....           0.792434   
12  {'rfc__max_depth': 22, 'rfc__max_features': 'a...           0.788933   
13  {'rfc__max_depth': 22, 'rfc__max_features': 'a...           0.791078   
14  {'rfc__max_depth': 22, 'rfc__max_features': 's...           0.788933   
15  {'rfc__max_depth': 22, 'rfc__max_features': 's...           0.791

### XGBoost
